lorawan-device/src/async_device/mod.rs

```diff
@@ -254,7 +281,64 @@ where
     /// `Response::DownlinkReceived` is returned from `send`. This call consumes the downlink
     /// data. If no downlink data is available, `None` is returned.
     pub fn take_downlink(&mut self) -> Option<Downlink> {
-        self.downlink.take()
+        self.downlink.pop()
+    }
+
+    async fn window_complete(&mut self) -> Result<(), Error<R::PhyError>> {
+        if self.class_c {
+            let rf_config = self.mac.region.get_rxc_config(self.mac.configuration.data_rate);
+            self.radio.setup_rx(rf_config).await.map_err(Error::Radio)?;
+        } else {
+            self.radio.low_power().await.map_err(Error::Radio)?;
+        }
+        Ok(())
+    }
+
+    async fn between_windows(
+        &mut self,
+        duration: u32,
+    ) -> Result<Option<mac::Response>, Error<R::PhyError>> {
+        if self.class_c {
```

Comment:
Could you switch the check to deal with `if !self.class_c` as this would allow to do an early return and reduce the indent level for the "big part" of this function...
Once code for class B arrives, this "flag" could be turned to optional enum (None, Some(class_b), Some(class_c)...

Reply:
Thanks, I like the idea! Not that it matters, but I'm skeptical that Class B will ever play into this section though.